### PR TITLE
Copyover from cockpit

### PR DESCRIPTION
**Pull request by Gundersanne**
So the result of a tests-scan for this project looks something like this:

```
INFO:root:Processing #1 titled 'WIP: Copyover from cockpit' on revision 4552d986ec9861e74dd1d34ff65db0fee2ed3eaf
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30/container-bastion ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30/selenium-firefox ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30/selenium-chrome ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  debian-stable ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  debian-testing ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  ubuntu-1804 ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  ubuntu-stable ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30 ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-atomic ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/starter-kit --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/starter-kit TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  centos-7 ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/starter-kit --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/starter-kit TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30 ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit-ostree --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit-ostree TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-atomic ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit-ostree --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit-ostree TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  continuous-atomic ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit-podman --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit-podman TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-29 ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit-podman --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=cockpit-project/cockpit-podman TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30 ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=weldr/lorax --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=weldr/lorax TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30 ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=weldr/lorax --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=weldr/lorax TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30/tar ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=weldr/lorax --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=weldr/lorax TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30/live-iso ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=weldr/lorax --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=weldr/lorax TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30/qcow2 ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=weldr/lorax --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=weldr/lorax TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30/aws ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=weldr/lorax --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=weldr/lorax TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30/azure ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=weldr/cockpit-composer --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=weldr/cockpit-composer TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30/chrome ; cd ../..
PRIORITY=0004 bots/make-checkout --verbose --base=master --repo=weldr/cockpit-composer --bots-ref=pull/1/head master && cd bots/make-checkout-workdir && GITHUB_BASE=cockpit-project/bots TEST_PROJECT=weldr/cockpit-composer TEST_NAME=pull-1-20190829-150101 TEST_REVISION=4552d986ec9861e74dd1d34ff65db0fee2ed3eaf bots/tests-invoke --pull-number=1  fedora-30/firefox ; cd ../..
```

### CODE REVIEW

File: make-checkout [hunk 1]
```diff
@@ -70,11 +70,24 @@ def main():
...
+    if api.repo == "cockpit-project/cockpit":
+        if opts.base and opts.base != "master":
+            execute("git", "remote", "add", "bots_origin", "https://github.com/cockpit-project/bots", cwd=TARGET_DIR)
+            execute("git", "fetch", "--depth=1", "bots_origin", opts.bots_ref or "master", cwd=TARGET_DIR)
+            bots_ref = execute("git", "rev-parse", "FETCH_HEAD", cwd=TARGET_DIR).strip()
+            execute("git", "checkout", "--force", bots_ref or "bots_origin/master", "--", "bots/", cwd=TARGET_DIR, error_on_fail=False)
+            execute("git", "reset", "bots", cwd=TARGET_DIR)
+
+    else:
+        execute("git", "clone", "--reference-if-able", "{}/{}".format(cache, "cockpit-project/bots"), "https://github.com/cockpit-project/bots", bots_clone_dir)
```
Comment (martinpitt) [inline make-checkout]: This is already being run from a master clone of cockpit-project/bots, right? So why does this need to be fetched/checked out again into some cache dir? Couldn't this just always clone itself? (I guess the point of this new clone is to not destroy your own bots master dir with --bots-ref?)

File: .gitignore [hunk 2]
```diff
@@ -6,3 +6,4 @@
 /*.log
 /build-results/
 make-checkout-workdir
+bots
```
Comment (martinpitt) [inline .gitignore]: Why is that necessary? The extra clone of bots should always be in make-checkout-workdir/bots, not in ./bots?

File: make-checkout [hunk 3]
```diff
@@ -70,11 +70,24 @@ def main():
...
-        execute("git", "remote", "add", "bots_origin", "https://github.com/cockpit-project/cockpit", cwd=TARGET_DIR)
-        execute("git", "fetch", "--depth=1", "bots_origin", opts.bots_ref or "master", cwd=TARGET_DIR)
-        bots_ref = execute("git", "rev-parse", "FETCH_HEAD", cwd=TARGET_DIR).strip()
-        execute("git", "checkout", "--force", bots_ref or "bots_origin/master", "--", "bots/", cwd=TARGET_DIR, error_on_fail=False)
+    bots_clone_dir = os.path.join(TARGET_DIR, "bots")
+
+    # HACK
+    # As long as bots/ is duplicated in cockpit, keep the conditional section here
+    if api.repo == "cockpit-project/cockpit":
+        if opts.base and opts.base != "master":
```
Comment (martinpitt) [inline make-checkout]: At first sigh the original condition was correct -- you also need to run the "else:" parts for non-master cockpit branches, as they also don't have bots/. However, I think this special case should  look differently: If you clone cockpit/master, it should remove its bots/ directory afterwards. This makes this whole code part much easier to understand, and will run through the bots/ setup the same way for every PR.

File: make-checkout [hunk 4]
```diff
@@ -70,11 +70,24 @@ def main():
...
+            execute("git", "remote", "add", "bots_origin", "https://github.com/cockpit-project/bots", cwd=TARGET_DIR)
```
Comment (martinpitt) [inline make-checkout]: As long as the One True Bots directory is still in cockpit, let's please use that, not cockpit-project/bots. We can flip this over once we declare cockpit-project/bots as The Truth.

For PRs against cockpit-project/cockpit master, this plus the special case above will essentially mean to remove bots/ and replace it with an identical copy. This should be fine, and the extra overhead is only temporary.

File: .gitignore (hunk 2 above)
Comment (Gundersanne) [reply .gitignore]: The commands outputted by tests-scan is `bots/make-checkout`. But relative to cockpit-project/bots it should be `./make-checkout`. During trying to simulate this I found it easier to just create a bots/ within bots.

But maybe it's worth changing tests-scan so the make-checkout command is relative to tests-scan.

File: make-checkout (hunk 1 above)
Comment (Gundersanne) [reply make-checkout]: Well I guess we could clone it from itself, and then execute the fetch of bots_ref within that clone.

So like `git clone . make-checkout-workdir/bots` etc...

File: make-checkout (hunk 3 above)
Comment (Gundersanne) [reply make-checkout]: Ack

File: make-checkout (hunk 1 above)
Comment (Gundersanne) [reply make-checkout]: Though if we want to have cockpit-project/cockpit still be the source of truth, we don't really need to do a clone, we can just use the same thing we've been using; a fetch --depth1 and force checkout.

File: .gitignore (hunk 2 above)
Comment (martinpitt) [reply .gitignore]: Thanks, let's live with the bots/bots thing until bots/ gets dropped from cockpit, and we can clean up the temporary hacks.
